File: sdks/typescript/src/adapters/memory.ts
/**
 * In-Memory Adapter for Cognitive Memory
 *
 * Three-tier dict-based storage (hot/cold/stub) with brute-force cosine similarity.
 * Zero dependencies, great for testing and single-process use.
 */

import { randomUUID } from "node:crypto";
import type { Memory, ScoredMemory } from "../core/types";
import { createDefaultMemory } from "../core/types";
import { cosineSimilarity } from "../utils/embeddings";
import { MemoryAdapter, type MemoryFilters } from "./base";

export class InMemoryAdapter extends MemoryAdapter {
  /** Hot store — active memories */
  hot = new Map<string, Memory>();
  /** Cold store — inactive but retained */
  cold = new Map<string, Memory>();
  /** Stub store — archived summaries */
  stubs = new Map<string, Memory>();

  private links = new Map<
    string,
    { strength: number; createdAt: number; updatedAt: number }
  >();
  private now: () => number;
  private idFactory: () => string;

  constructor(options?: { now?: () => number; idFactory?: () => string }) {
    super();
    this.now = options?.now ?? Date.now;
    this.idFactory = options?.idFactory ?? randomUUID;
  }

  // ------------------------------------------------------------------
  // CRUD
  // ------------------------------------------------------------------

  async transaction<T>(
    callback: (adapter: MemoryAdapter) => Promise<T>,
  ): Promise<T> {
    return callback(this);
  }

  async createMemory(
    memory: Omit<Memory, "id" | "createdAt" | "updatedAt">,
  ): Promise<string> {
    const id = this.idFactory();
    const now = this.now();
    const m = createDefaultMemory({
      ...memory,
      id,
      createdAt: now,
      updatedAt: now,
    });
    this.hot.set(id, m);
    return id;
  }

  async getMemory(id: string): Promise<Memory | null> {
    return this.hot.get(id) ?? this.cold.get(id) ?? this.stubs.get(id) ?? null;
  }

  async getMemories(ids: string[]): Promise<Memory[]> {
    return ids
      .map((id) => this.hot.get(id) ?? this.cold.get(id) ?? this.stubs.get(id))
      .filter((m): m is Memory => m !== undefined);
  }

  async queryMemories(filters: MemoryFilters): Promise<Memory[]> {
    let items = [
      ...Array.from(this.hot.values()),
      ...Array.from(this.cold.values()),
    ];

    // Exclude superseded by default
    if (!filters.includeSuperseded) {
      items = items.filter((m) => !m.isSuperseded);
    }

    // Exclude stubs from general queries
    items = items.filter((m) => !m.isStub);

    if (filters.userId)
      items = items.filter((m) => m.userId === filters.userId);
    if (filters.categories)
      items = items.filter((m) => filters.categories!.includes(m.category));
    if (filters.memoryTypes)
      items = items.filter((m) => filters.memoryTypes!.includes(m.memoryType));
    if (filters.minRetention !== undefined)
      items = items.filter((m) => m.retention >= filters.minRetention!);
    if (filters.minImportance !== undefined)
      items = items.filter((m) => m.importance >= filters.minImportance!);
    if (filters.createdAfter !== undefined)
      items = items.filter((m) => m.createdAt >= filters.createdAfter!);
    if (filters.createdBefore !== undefined)
      items = items.filter((m) => m.createdAt <= filters.createdBefore!);
    if (filters.offset) items = items.slice(filters.offset);
    if (filters.limit) items = items.slice(0, filters.limit);
    return items;
  }

  async updateMemory(id: string, updates: Partial<Memory>): Promise<void> {
    const tier = this.getTier(id);
    if (!tier) return;
    const existing = tier.get(id)!;
    tier.set(id, {
      ...existing,
      ...updates,
      id,
      createdAt: existing.createdAt,
    });
  }

  async deleteMemory(id: string): Promise<void> {
    this.hot.delete(id);
    this.cold.delete(id);
    this.stubs.delete(id);
  }

  async deleteMemories(ids: string[]): Promise<void> {
    for (const id of ids) {
      this.hot.delete(id);
      this.cold.delete(id);
      this.stubs.delete(id);
    }
  }

  // ------------------------------------------------------------------
  // Vector search
  // ------------------------------------------------------------------

  async vectorSearch(
    embedding: number[],
    filters?: MemoryFilters,
  ): Promise<ScoredMemory[]> {
    let items = Array.from(this.hot.values());

    // Include cold store memories in search if needed
    if (filters?.includeSuperseded) {
      items = [...items, ...Array.from(this.cold.values())];
    }

    // Apply filters
    if (filters?.userId)
      items = items.filter((m) => m.userId === filters.userId);
    if (filters?.categories)
      items = items.filter((m) => filters.categories!.includes(m.category));
    if (filters?.memoryTypes)
      items = items.filter((m) => filters.memoryTypes!.includes(m.memoryType));
    if (filters?.minRetention !== undefined)
      items = items.filter((m) => m.retention >= filters.minRetention!);
    if (!filters?.includeSuperseded)
      items = items.filter((m) => !m.isSuperseded);
    items = items.filter((m) => !m.isStub);

    return items
      .map((m) => {
        const relevanceScore = cosineSimilarity(embedding, m.embedding);
        return {
          ...m,
          relevanceScore,
          finalScore: relevanceScore * m.retention,
        };
      })
      .sort((a, b) => b.relevanceScore - a.relevanceScore)
      .slice(0, filters?.limit ?? 5);
  }

  // ------------------------------------------------------------------
  // Retention
  // ------------------------------------------------------------------

  async updateRetentionScores(updates: Map<string, number>): Promise<void> {
    for (const [id, retention] of updates.entries()) {
      const tier = this.getTier(id);
      if (!tier) continue;
      const m = tier.get(id)!;
      tier.set(id, { ...m, retention });
    }
  }

  // ------------------------------------------------------------------
  // Links
  // ------------------------------------------------------------------

  async createOrStrengthenLink(
    sourceId: string,
    targetId: string,
    strength: number,
  ): Promise<void> {
    const key = this.linkKey(sourceId, targetId);
    const existing = this.links.get(key);
    const now = this.now();
    this.links.set(key, {
      strength: Math.min(1, (existing?.strength ?? 0) + strength),
      createdAt: existing?.createdAt ?? now,
      updatedAt: now,
    });
  }

  async getLinkedMemories(
    memoryId: string,
    minStrength: number = 0.3,
  ): Promise<Array<Memory & { linkStrength: number }>> {
    return this.getLinkedMemoriesMultiple([memoryId], minStrength);
  }

  async getLinkedMemoriesMultiple(
    memoryIds: string[],
    minStrength: number = 0.3,
  ): Promise<Array<Memory & { linkStrength: number }>> {
    const out = new Map<string, Memory & { linkStrength: number }>();
    for (const id of memoryIds) {
      for (const [key, row] of this.links.entries()) {
        if (row.strength < minStrength) continue;
        const [a, b] = key.split("|");
        const other = a === id ? b : b === id ? a : null;
        if (!other) continue;
        const m = await this.getMemory(other);
        if (!m) continue;
        const prev = out.get(other);
        out.set(
          other,
          prev
            ? { ...m, linkStrength: Math.max(prev.linkStrength, row.strength) }
            : { ...m, linkStrength: row.strength },
        );
      }
    }
    return Array.from(out.values());
  }

  async deleteLink(sourceId: string, targetId: string): Promise<void> {
    this.links.delete(this.linkKey(sourceId, targetId));
  }

  // ------------------------------------------------------------------
  // Consolidation helpers
  // ------------------------------------------------------------------

  async findFadingMemories(
    userId: string,
    maxRetention: number,
  ): Promise<Memory[]> {
    return [
      ...Array.from(this.hot.values()),
      ...Array.from(this.cold.values()),
    ].filter(
      (m) =>
        m.userId === userId &&
        m.retention < maxRetention &&
        !m.isSuperseded &&
        !m.isStub,
    );
  }

  async findStableMemories(
    userId: string,
    minStability: number,
    minAccessCount: number,
  ): Promise<Memory[]> {
    return Array.from(this.hot.values()).filter(
      (m) =>
        m.userId === userId &&
        m.stability >= minStability &&
        m.accessCount >= minAccessCount,
    );
  }

  async markSuperseded(memoryIds: string[], summaryId: string): Promise<void> {
    for (const id of memoryIds) {
      const tier = this.getTier(id);
      if (!tier) continue;
      const m = tier.get(id)!;
      tier.set(id, {
        ...m,
        isSuperseded: true,
        supersededBy: summaryId,
        metadata: { ...(m.metadata ?? {}), supersededBy: summaryId },
      });
    }
  }

  // ------------------------------------------------------------------
  // Tiered storage
  // ------------------------------------------------------------------

  async migrateToCold(memoryId: string, coldSince: number): Promise<void> {
    const m = this.hot.get(memoryId);
    if (!m) return;
    this.hot.delete(memoryId);
    this.cold.set(memoryId, {
      ...m,
      isCold: true,
      coldSince,
    });
  }

  async migrateToHot(memoryId: string): Promise<void> {
    const m = this.cold.get(memoryId);
    if (!m) return;
    this.cold.delete(memoryId);
    this.hot.set(memoryId, {
      ...m,
      isCold: false,
      coldSince: null,
      daysAtFloor: 0,
    });
  }

  async convertToStub(memoryId: string, stubContent: string): Promise<void> {
    const m = this.cold.get(memoryId) ?? this.hot.get(memoryId);
    if (!m) return;
    this.hot.delete(memoryId);
    this.cold.delete(memoryId);
    this.stubs.set(memoryId, {
      ...m,
      content: stubContent,
      isStub: true,
      isCold: false,
      coldSince: null,
      embedding: [],
    });
  }

  // ------------------------------------------------------------------
  // Traversal
  // ------------------------------------------------------------------

  async allActive(): Promise<Memory[]> {
    return [
      ...Array.from(this.hot.values()),
      ...Array.from(this.cold.values()),
    ];
  }

  async allHot(): Promise<Memory[]> {
    return Array.from(this.hot.values());
  }

  async allCold(): Promise<Memory[]> {
    return Array.from(this.cold.values());
  }

  // ------------------------------------------------------------------
  // Counts
  // ------------------------------------------------------------------

  async hotCount(): Promise<number> {
    return this.hot.size;
  }

  async coldCount(): Promise<number> {
    return this.cold.size;
  }

  async stubCount(): Promise<number> {
    return this.stubs.size;
  }

  async totalCount(): Promise<number> {
    return this.hot.size + this.cold.size + this.stubs.size;
  }

  // ------------------------------------------------------------------
  // Reset
  // ------------------------------------------------------------------

  async clear(): Promise<void> {
    this.hot.clear();
    this.cold.clear();
    this.stubs.clear();
    this.links.clear();
  }

  // ------------------------------------------------------------------
  // Private helpers
  // ------------------------------------------------------------------

  private linkKey(a: string, b: string): string {
    return a < b ? `${a}|${b}` : `${b}|${a}`;
  }

  private getTier(id: string): Map<string, Memory> | null {
    if (this.hot.has(id)) return this.hot;
    if (this.cold.has(id)) return this.cold;
    if (this.stubs.has(id)) return this.stubs;
    return null;
  }
}

File: sdks/python/src/cognitive_memory/adapters/memory.py
"""
In-memory adapter for cognitive-memory.

Tiered storage with hot/cold/stub dicts and brute-force cosine similarity.
Zero dependencies, great for testing and single-process use.
"""

from __future__ import annotations

import math
from collections import Counter
from datetime import datetime
from typing import Optional, Callable, TypeVar

from ..types import Memory, MemoryCategory
from ..embeddings import cosine_similarity
from .base import MemoryAdapter

T = TypeVar("T")


class InMemoryAdapter(MemoryAdapter):
    """
    Dict-based tiered in-memory store.

    Hot: vector-indexed, supports similarity search.
    Cold: key-value only, accessed by ID.
    Stub: lightweight summary records, no embedding.
    """

    def __init__(self):
        self.hot: dict[str, Memory] = {}
        self.cold: dict[str, Memory] = {}
        self.stubs: dict[str, Memory] = {}

    # ------------------------------------------------------------------
    # CRUD
    # ------------------------------------------------------------------

    async def create(self, memory: Memory) -> None:
        self.hot[memory.id] = memory

    async def get(self, memory_id: str) -> Optional[Memory]:
        if memory_id in self.hot:
            return self.hot[memory_id]
        if memory_id in self.cold:
            return self.cold[memory_id]
        if memory_id in self.stubs:
            return self.stubs[memory_id]
        return None

    async def get_batch(self, memory_ids: list[str]) -> list[Memory]:
        results = []
        for mid in memory_ids:
            m = await self.get(mid)
            if m is not None:
                results.append(m)
        return results

    async def update(self, memory: Memory) -> None:
        if memory.id in self.hot:
            self.hot[memory.id] = memory
        elif memory.id in self.cold:
            self.cold[memory.id] = memory
        elif memory.id in self.stubs:
            self.stubs[memory.id] = memory

    async def delete(self, memory_id: str) -> None:
        self.hot.pop(memory_id, None)
        self.cold.pop(memory_id, None)
        self.stubs.pop(memory_id, None)

    async def delete_batch(self, memory_ids: list[str]) -> None:
        for mid in memory_ids:
            await self.delete(mid)

    # ------------------------------------------------------------------
    # Vector search
    # ------------------------------------------------------------------

    async def search_similar(
        self,
        query_embedding: list[float],
        top_k: int = 10,
        include_superseded: bool = False,
    ) -> list[tuple[Memory, float]]:
        results = []
        for mem in self.hot.values():
            if mem.embedding is None:
                continue
            if mem.is_superseded and not include_superseded:
                continue
            if mem.is_stub:
                continue
            sim = cosine_similarity(query_embedding, mem.embedding)
            results.append((mem, sim))

        results.sort(key=lambda x: x[1], reverse=True)
        return results[:top_k]

    # ------------------------------------------------------------------
    # Lexical search (BM25)
    # ------------------------------------------------------------------

    @staticmethod
    def _tokenize(text: str) -> list[str]:
        return text.lower().split()

    async def search_lexical(
        self,
        query: str,
        top_k: int = 10,
        include_superseded: bool = False,
    ) -> list[tuple[Memory, float]]:
        """BM25 lexical search over hot store."""
        query_tokens = self._tokenize(query)
        if not query_tokens:
            return []

        # Build corpus stats
        docs: list[tuple[Memory, list[str]]] = []
        for mem in self.hot.values():
            if mem.is_stub:
                continue
            if mem.is_superseded and not include_superseded:
                continue
            docs.append((mem, self._tokenize(mem.content)))

        if not docs:
            return []

        N = len(docs)
        avgdl = sum(len(tokens) for _, tokens in docs) / N if N else 1.0
        k1, b = 1.2, 0.75

        # IDF per query token
        df: dict[str, int] = Counter()
        for _, tokens in docs:
            unique = set(tokens)
            for t in query_tokens:
                if t in unique:
                    df[t] += 1

        idf: dict[str, float] = {}
        for t in query_tokens:
            n = df.get(t, 0)
            idf[t] = math.log((N - n + 0.5) / (n + 0.5) + 1.0)

        # Score each doc
        results: list[tuple[Memory, float]] = []
        for mem, tokens in docs:
            tf = Counter(tokens)
            dl = len(tokens)
            score = 0.0
            for t in query_tokens:
                f = tf.get(t, 0)
                if f == 0:
                    continue
                num = f * (k1 + 1)
                denom = f + k1 * (1 - b + b * dl / avgdl)
                score += idf.get(t, 0.0) * num / denom
            if score > 0:
                results.append((mem, score))

        results.sort(key=lambda x: x[1], reverse=True)
        return results[:top_k]

    # ------------------------------------------------------------------
    # Tiered storage
    # ------------------------------------------------------------------

    async def migrate_to_cold(self, memory_id: str, cold_since: datetime) -> None:
        if memory_id not in self.hot:
            return
        mem = self.hot.pop(memory_id)
        mem.is_cold = True
        mem.cold_since = cold_since
        self.cold[memory_id] = mem

    async def migrate_to_hot(self, memory_id: str) -> None:
        if memory_id not in self.cold:
            return
        mem = self.cold.pop(memory_id)
        mem.is_cold = False
        mem.cold_since = None
        mem.days_at_floor = 0
        self.hot[memory_id] = mem

    async def convert_to_stub(self, memory_id: str, stub_content: str) -> None:
        if memory_id in self.cold:
            old = self.cold.pop(memory_id)
        elif memory_id in self.hot:
            old = self.hot.pop(memory_id)
        else:
            return

        stub = Memory(
            id=old.id,
            content=stub_content,
            category=old.category,
            importance=old.importance,
            stability=0.0,
            created_at=old.created_at,
            is_stub=True,
            is_cold=True,
            embedding=None,
        )
        self.stubs[memory_id] = stub

    # ------------------------------------------------------------------
    # Associative links
    # ------------------------------------------------------------------

    async def create_or_strengthen_link(
        self, source_id: str, target_id: str, weight: float,
    ) -> None:
        # Links are stored directly on Memory objects in the in-memory adapter
        # This is a no-op — the engine manages links on Memory.associations directly
        pass

    async def get_linked_memories(
        self, memory_id: str, min_weight: float = 0.3,
    ) -> list[tuple[Memory, float]]:
        mem = await self.get(memory_id)
        if mem is None:
            return []
        results = []
        for assoc in mem.associations.values():
            if assoc.weight < min_weight:
                continue
            target = await self.get(assoc.target_id)
            if target is not None and not target.is_stub:
                results.append((target, assoc.weight))
        return results

    async def delete_link(self, source_id: str, target_id: str) -> None:
        source = await self.get(source_id)
        if source and target_id in source.associations:
            del source.associations[target_id]
        target = await self.get(target_id)
        if target and source_id in target.associations:
            del target.associations[source_id]

    # ------------------------------------------------------------------
    # Consolidation helpers
    # ------------------------------------------------------------------

    async def find_fading(
        self, threshold: float, exclude_core: bool = True,
    ) -> list[Memory]:
        # Note: caller must compute retention separately — this is a storage-level
        # filter. For in-memory adapter, we return all hot non-superseded memories
        # and let the engine filter by computed retention.
        results = []
        for mem in self.hot.values():
            if mem.is_superseded:
                continue
            if exclude_core and mem.category == MemoryCategory.CORE:
                continue
            results.append(mem)
        return results

    async def find_stable(
        self, min_stability: float, min_access_count: int,
    ) -> list[Memory]:
        results = []
        for mem in list(self.hot.values()) + list(self.cold.values()):
            if mem.stability >= min_stability and mem.access_count >= min_access_count:
                results.append(mem)
        return results

    async def mark_superseded(
        self, memory_ids: list[str], summary_id: str,
    ) -> None:
        for mid in memory_ids:
            mem = await self.get(mid)
            if mem is not None:
                mem.is_superseded = True
                mem.superseded_by = summary_id

    # ------------------------------------------------------------------
    # Traversal
    # ------------------------------------------------------------------

    async def all_active(self) -> list[Memory]:
        return list(self.hot.values()) + list(self.cold.values())

    async def all_hot(self) -> list[Memory]:
        return list(self.hot.values())

    async def all_cold(self) -> list[Memory]:
        return list(self.cold.values())

    # ------------------------------------------------------------------
    # Counts
    # ------------------------------------------------------------------

    async def hot_count(self) -> int:
        return len(self.hot)

    async def cold_count(self) -> int:
        return len(self.cold)

    async def stub_count(self) -> int:
        return len(self.stubs)

    async def total_count(self) -> int:
        return len(self.hot) + len(self.cold) + len(self.stubs)

    # ------------------------------------------------------------------
    # Batch operations
    # ------------------------------------------------------------------

    async def batch_update(self, memories: list[Memory]) -> None:
        for mem in memories:
            await self.update(mem)

    async def update_retention_scores(self, updates: dict[str, float]) -> None:
        # Retention is computed on-the-fly in the in-memory adapter,
        # so this is a no-op. Remote adapters cache retention.
        pass

    # ------------------------------------------------------------------
    # Transactions
    # ------------------------------------------------------------------

    async def transaction(self, callback):
        # No-op for in-memory — just call directly
        return await callback(self)

    # ------------------------------------------------------------------
    # Reset
    # ------------------------------------------------------------------

    async def clear(self) -> None:
        self.hot.clear()
        self.cold.clear()
        self.stubs.clear()
